src/librustc_mir/transform/generator.rs

```diff
@@ -539,9 +539,23 @@ fn locals_live_across_suspend_points(
             let mut live_locals_here = storage_required;
             live_locals_here.intersect(&liveness.outs[block]);
 
-            // The generator argument is ignored
+            // The generator argument is ignored.
             live_locals_here.remove(self_arg());
 
+            // If the resume arg ends up in a plain local, we ignore the local's liveness in the
```

Comment:
This assumption doesn't hold in cases where the resume arg is `Drop`. If the generator is dropped before resuming, we will need to read its value in the destructor.

In the remaining cases, it would be correct to say that the local isn't live across the yield point.

In fact, the existing liveness analysis should work if it considers the yield to be a def of its resume arg. Yields with resume args are already [treated as stores](https://github.com/rust-lang/rust/blob/75cf41afb468152611212271bae026948cd3ba46/src/librustc/mir/visit.rs#L516-L528) on the resume arg, and it does the [right thing](https://github.com/rust-lang/rust/blob/75cf41afb468152611212271bae026948cd3ba46/src/librustc/mir/visit.rs#L918-L924) for projections. So at first glance, it seems like we shouldn't need this change at all.

The reason it doesn't work today probably has to do with the liveness algorithm and where it "places" the effects (i.e. it is probably acting like a write occurs _before_ the yield instead of after it). We may need something analogous to the dataflow framework's `call_return_effect` here.

Reply:
Yeah, my first attempt at fixing this tried to do exactly that (including adding a `resume_effect`), but it ended up just reintroducing #69039, so I took this simpler approach instead. I don't think I currently understand all the dataflow analyses that feed into the generator transform well enough, but I can look into this again if you want.

Reply:
I wasn't imagining doing this in dataflow, but in the [liveness analysis](https://github.com/rust-lang/rust/blob/master/src/librustc_mir/util/liveness.rs). The yield should be modeled as a def that occurs on resume.

(Though now that I think about it, we need it in dataflow too, to model the assignment from the yield.)

Reply:
If I'm not mistaken that should already happen due to the `MutatingUse` classification. The issue with that is that `liveness` isn't using the dataflow framework for some reason, so it has no way to model events that only happen along certain edges in the CFG.

I don't entirely understand why we're using what essentially boils down to two liveness analyses here though, it seems to me that `MaybeRequiresStorage` is strictly more precise than `liveness` since it uses the dataflow framework? The generator transform requires that *both* analyses say a local is live anyways.

Reply:
Ah, I think I understand this a bit better now: `liveness` is indeed an actual liveness analysis that tracks variables whose values may be used later. `MaybeRequiresStorage` ignores whether locals are actually going to be used later but tries to determine a more accurate live range of locals than the existing `StorageLive` and `StorageDead` markers.

Reply:
Yep, that's right. Sorry the code isn't more clear. (Feel free to ping me on zulip or discord if you get confused -- I'm usually more responsive on those.)

Reply:
Okay, good. So, do you think this will need any changes to liveness? AFAICT liveness handles `Yield` correctly since f2c1468965a7af5887d353adf77427344327be0d (if it didn't, the resume argument wouldn't be saved at all, which means things would break horribly).

Reply:
My thinking is that if the liveness analysis is modeling everything correctly, we would never have needed this code here. The resume arg local should not have been considered live across the yield point, because the write to it happens immediately _after_ the yield, not before. If a write to a variable occurs immediately after a yield point, it can't be considered live across that point. (Unless that variable is `Drop`, in which case it would be live because of the drop glue).

It looks like you've accomplished the same effect by modeling this in dataflow instead of liveness, but I don't think it will be as effective in all cases.

For example, let's assume that `x` is not `Drop`. If I put two `x = yield;` statements in a row and `x`, the generator size shouldn't change (we still never use an earlier value of x after resuming from a yield point). I would guess that currently the size does change, because we're relying on the storage-required analysis. Storage-required gets reset on `drop(x);`, but not on overwrite.

I think these dataflow changes are good on their own and make some cases work better than they did before, but I would like to make similar changes for liveness. If you'd rather we merge this now and do the remaining work in another PR, I'm fine with that.

Reply:
Ah, I see. That makes sense, but it seems like a relatively big refactor (since it requires changing `liveness` to use the dataflow framework), so I'd prefer to do it later.

Reply:
Just catching up now.

My understanding is that `MaybeRequiresStorage` could basically be `(liveness | maybe_borrowed) & storage_live`, where `liveness` is the [canonical backwards dataflow analysis](https://en.wikipedia.org/wiki/Live_variable_analysis) and `storage_live` uses `StorageLive`/`StorageDead` to determine whether borrows of certain locals have been invalidated. Is this correct?

The changes to the dataflow framework are good (I might rename some stuff in a cleanup PR, but that's just my OCD).  However, I think it is worthwhile to try to express `MaybeRequiresStorage` as a combination of component analyses. Whether this PR should be merged before that attempt is made is up to y'all, since it depends on the magnitude of the memory savings.

One clarification, the pass in `utils/liveness.rs` is still a dataflow pass, just not a forwards one. I can extend the framework to also handle backwards dataflow, but since generator storage is the only place that analysis is used, you should feel free to make changes to it directly. It's not a major refactor.

Reply:
Ah okay, I didn't realize it's the only backwards analysis. This PR isn't exactly urgent, but it's somewhat of a soft blocker for https://github.com/rust-lang/rust/pull/69033, which I'd love to have available in nightly as soon as that seems reasonable. Thanks!

Reply:
I don't think liveness analysis would use the dataflow framework actually. As I recall, it's a fundamentally different computation. We'd need to extend what's there to support the same kind of success effects that dataflow supports, though.

Reply:
@ecstatic-morse it might be possible to express that, but there's some extra stuff that `MaybeRequiresStorage` does today. If a value is moved from and it hasn't been borrowed, then it doesn't require storage. I'm not sure where to stick that logic in a refactor (if it goes in `maybe_borrowed` but not `liveness` for example, we'll optimize fewer cases).

Also fwiw, I believe @eddyb's previous work on NRVO involved bidirectional dataflow, so it might be worth introducing at some point :)

Reply:
You can think of a move from a local as `x = undef` for the purposes of the liveness analysis, i.e. `x` is "kill"-ed (no longer live) when it is moved out of.